fix(orderBy): ensure correct ordering with arrays of objects and no predicate

petebacondarwin opened the pull request:
This fix also considerably refactors how the `orderBy` filter works in internally.

The new algorithm precomputes the predicate values for the array being ordered. This makes the algorith easier to follow but also ensures that this computation is done a maximum of n times, whereas in the previous algorithm it could be greater. The downside is that we must temporarily store this intermediate array.

There could be performance implications, but they are as likely to be beneficial as detrimental. It could probably do with benchmarking.

Closes #11866

petebacondarwin commented:
@Narretz and wawyed could you take a look?

petebacondarwin commented:
I guess this should be backported to 1.3.x - but not to 1.2.x since it uses `Array.map` and `Array.reduce`, which are not available in IE8.

petebacondarwin commented:
Here is a plunker with this build of AngularJS demonstrating the original bug is fixed: http://plnkr.co/edit/WxAW671zu7BePcN4wgBn?p=preview

Narretz commented on test/ng/filter/orderBySpec.js:
```diff
@@ -11,15 +11,15 @@ describe('Filter: orderBy', function() {
-      expect(orderBy([2, 1, 3], [''])).toEqual([1, 2, 3]);
+      // expect(orderBy([2, 1, 3], '')).toEqual([1, 2, 3]);
+      // expect(orderBy([2, 1, 3], [])).toEqual([1, 2, 3]);
+      // expect(orderBy([2, 1, 3], [''])).toEqual([1, 2, 3]);
```
Is this intentionally commented out?

petebacondarwin replied:
Oops. No

wawyed commented:
Seems to be working fine when setting reverse true, although it doesn't seem to be working when using the '-' notation instead of reverse true for array with objects.

Pulnker with the issue:

http://plnkr.co/edit/Z3cdF3wMJ7Uzubw2Gma9?p=preview

petebacondarwin commented:
Agh! I knew it wasn't going to be so easy.
Now here is the problem. If you are reversing with a predicate then you need a way of telling the algorithm that objects should be ordered by index, rather than identity - but only if they don't have a `toString` or `toValue` or are being mapped by specific predicate.

petebacondarwin commented:
OK, so I fixed it...

Actually, the reason for the refactoring was that it made this scenario much easier to handle.

But I am not 100% happy that we have to guess at whether the object is a POJO or not, when we decide to use the index rather than the identity for its ordering value.

Also, we now treat `null` as lower order than an object - hence the change to two of the tests. I don't think this should really affect anyone but perhaps it should be marked as a potential breaking change?

petebacondarwin commented:
Agh! So safari wasn't happy comparing nulls, so I have now fixed that too

petebacondarwin commented:
http://plnkr.co/edit/BSkUBBUfg0c9TJ9tTK5F?p=preview

wawyed commented:
For example in this plunkr is this behaviour desired for the third ng-repeat?:

http://plnkr.co/edit/gyw8B10H9tpGdu77g1e7?p=preview

According to documentation if reverse is true it should reverse the order of the array, no matter what type the contents are. It's not that clear what it should do with '-'.

lgalfaso commented on src/ng/filter/orderBy.js:
```diff
@@ -176,88 +176,110 @@
-    sortPredicate = isArray(sortPredicate) ? sortPredicate : [sortPredicate];
+
+    if (!isArray(sortPredicate)) { sortPredicate = [sortPredicate]; }
     if (sortPredicate.length === 0) { sortPredicate = ['+']; }
```
Should this be
```js
sortPredicate.push('+');
```

?

lgalfaso replied:
Mmm... thinking a little more about this, I think my proposal and the current code are wrong as we are modifying the original `sortPredicate` and we should not do that

petebacondarwin replied:
I agree. We should not modify the passed in predicate

gkalpak replied:
How is the current implementation modifying the passed in predicate ?

petebacondarwin replied:
It's not. But I agree that we should start doing that either :-)

lgalfaso commented on src/ng/filter/orderBy.js:
```diff
@@ -176,88 +176,110 @@
-      if (isString(predicate)) {
+
+    var predicates = processPredicates(sortPredicate);
+    var compareValues = Array.prototype.map.call(array, getComparisonObject);
```
This is usually called *schwartzian transform*, a comment would help people understand what is going on

lgalfaso replied:
and please add Close #4282

lgalfaso commented on src/ng/filter/orderBy.js:
```diff
@@ -176,88 +176,110 @@
+
+    compareValues.sort(doComparison);
+    array = compareValues.map(function(item) { return item.value; });
+    if (reverseOrder) { array.reverse(); }
```
Can this be made part of the main comparison?

petebacondarwin replied:
It could but if you look at the previous implementation it meant complicating the processing of the predicate functions, which makes it significantly more messy.
What is the cost of `Array.reverse()`? It would seem to me that it could be done rather quickly inside the JavaScript engine. Any thoughts @caitp ?

petebacondarwin replied:
Actually, it is simple to move this into the predicates.

lgalfaso commented on src/ng/filter/orderBy.js:
```diff
@@ -176,88 +176,110 @@
+    }
+
+    function doComparison(v1, v2) {
+      return predicates.reduce(function(result, predicate, index) {
```
this may be a little too fancy just to avoid a for-loop

petebacondarwin replied:
I am not sure what is fancy about this. It is reducing a collection of comparison predicates down to a final comparison result.

petebacondarwin replied:
The for loop would look like:

```js
function doComparison(v1, v2) {
  var result = 0;
  for(var index=0, length = predicates.length; index<length; ++index) {
    result = result || compare(v1.predicateValues[index], v2.predicateValues[index]) * predicates[index].descending;
  }
  return result;
}
```

caitp replied:
honestly, for people who aren't familiar with `reduce`, this isn't significantly more readable, and is guaranteed to be slower. That and, comparison doesn't really map very well with the "reduce" operation

petebacondarwin replied:
@caitp The `reduce` is guaranteed to be slower?

caitp replied:
yes --- the spec mandates `HasProperty` checks for each item which are slow almost everywhere, and can't be optimized away because the callback can manipulate the array

caitp replied:
Basically for something that runs each digest on a potentially large collection, I'd prefer if it wasn't a reduce operation, both for readability and perf

petebacondarwin replied:
OK, I'll change this

lgalfaso commented:
> Also, we now treat null as lower order than an object - hence the change to two of the tests. I don't think this should really affect anyone but perhaps it should be marked as a potential breaking change?

Do you think it would make the code any harder to keep the original behavior? I do not mind this going one way or another, but do think this is a small breaking change

hamfastgamgee commented:
I'm guessing that this "small breaking change" might be accidentally fixing (or at least changing behavior of) #11312?

petebacondarwin commented:
I think we can probably move `null` in the sort order. I need to work out exactly where that is though.

petebacondarwin committed:
refact(filter): move `hasCustomToString` into shared namespace to be reused

petebacondarwin commented:
No breaking change no more: the algorithm sorts nulls in the same way as `array.sort()`.
Objects are now dealt with in the following manner:

* has a valid `toValue()` method: sort on the return value of this method
* has a valid custom `toString()` method (i.e. not `Object.prototype.toString()`): sort on the return value of this method
* otherwise sort on index of the object in the array

This means that dates are sorted as expected, but also that arrays are sorted by their `toString` representation, which is to convert each item to a string and then join them together, separated by commas. This gives the same result as calling `array.sort()` on such an array of arrays.

I have noted that the idiom used is a Schwartzian Transform and that it closes #4282.

lgalfaso commented on src/ng/filter/orderBy.js:
```diff
@@ -176,88 +176,110 @@
+
+    function doComparison(v1, v2) {
+      var result = 0;
+      for (var index=0, length = predicates.length; index < length; ++index) {
```
for (...; !result && index < length; ...) {

?

petebacondarwin replied:
I like

petebacondarwin replied:
But it is a bit obscure. I would prefer to make it more explicit:

```js
    function doComparison(v1, v2) {
      var result = 0;
      for (var index=0, length = predicates.length; index < length; ++index) {
        result = compare(v1.predicateValues[index], v2.predicateValues[index]) * predicates[index].descending;
        if (result) break;
      }
      return result;
    }
```

lgalfaso replied:
Works for me

lgalfaso commented on src/ng/filter/orderBy.js:
```diff
@@ -176,88 +176,110 @@
+      value = value.toString();
+      if (isPrimitive(value)) return value;
     }
+    return index;
```
trying to think when we end up returning `index` and what does it mean

petebacondarwin replied:
I'll add a comment. Is that enough or are you saying that the code might be wrong?

lgalfaso replied:
> In src/ng/filter/orderBy.js
> <https://github.com/angular/angular.js/pull/12072#discussion_r32614045>:
>
> >      }
> > +    return index;
>
> I'll add a comment. Is that enough or are you saying that the code might
> be wrong?
>

Before, we had that if the property was an object (that does not meet a
criteria) then the value was the blank string. Now it returns the position
in the index of the object.

This is, if you sort [{foo: 2}, {foo: 3}, {foo: 4}, {foo: {}}] by "foo",
over and over, then sorting it once will be different than sorting it twice
(assuming sort is stable)

petebacondarwin replied:
Well.... it is true that the sort will change on the second pass but I think it is reasonable to say that sorting mixed types like that is a rather strange case, no?

petebacondarwin replied:
Actually, this is not a problem. First of all the `orderBy` filter does not mutate the original array. Second if the types of things are different then they are ordered by their type first rather than their value.

lgalfaso replied:
I think this is very minor (famous last words), so there is no needs to keep this going on. Lets land this

petebacondarwin commented:
@lgalfaso I updated the code.

lgalfaso commented:
LGTM

petebacondarwin committed:
fix(orderBy): ensure correct ordering with arrays of objects and no predicate

By refactoring to use a Schwartzian transform, we can ensure that objects
with no custom `toString` or `toValue` methods are just ordered using
their position in the original collection.

Closes #11866
Closes #11312
Closes #4282
